src/main/scala/controllers/ActionHandler.scala

```diff
@@ -16,32 +17,40 @@ class ActionHandler @Inject()(applications: ApplicationOps, applicationForms: Ap
   import ApplicationData._
   import FieldCheckHelpers._
 
-
   def doSave(id: ApplicationId, sectionNumber: Int, fieldValues: JsObject): Future[Result] = {
-    sectionTypeFor(sectionNumber) match {
-      case VanillaSection =>
-        JsonHelpers.allFieldsEmpty(fieldValues) match {
-          case true => applications.deleteSection(id, sectionNumber).map(_ => redirectToOverview(id))
-          case false => applications.saveSection(id, sectionNumber, fieldValues).map(_ => redirectToOverview(id))
+    gatherSectionDetails(id, sectionNumber).flatMap {
+      case Some(app) =>
+        app.formSection.sectionType match {
+          case SectionTypeForm =>
+            JsonHelpers.allFieldsEmpty(fieldValues) match {
+              case true => applications.deleteSection(id, sectionNumber).map(_ => redirectToOverview(id))
```

Comment:
Well, I understand the consequence, but how can we be sure that there wouldn't be an opportunity where you can have an empty section being a completed section?